File: 백준/Gold/1744. 수 묶기/수 묶기.py
import sys
input = sys.stdin.readline

N = int(input())
p = []
m = []
for _ in range(N):
    n = int(input())
    if n > 0:
        p.append(n)
    else:
        m.append(n)

p.sort()
m.sort(reverse=True)

ans = 0

while len(p) >= 2:
    n1 = p.pop()
    n2 = p.pop()
    if n1 == 1 or n2 == 1:
        ans += n1 + n2
    else:
        ans += n1 * n2
if p:
    ans += p.pop()

while len(m) >= 2:
    n1 = m.pop()
    n2 = m.pop()
    ans += n1 * n2
if m:
    ans += m.pop()

print(ans)

File: 백준/Gold/1744. 수 묶기/수 묶기.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;
public class Main {
	public static void main(String[] args) throws IOException {
		
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		int N = Integer.parseInt(br.readLine());
		ArrayList<Long> p = new ArrayList<>();
		ArrayList<Long> m = new ArrayList<>();
		for (int i = 0; i < N; i++) {
			StringTokenizer st = new StringTokenizer(br.readLine());
			long n = Long.parseLong(st.nextToken());
			if (n > 0) {
				p.add(n);
			}
			else {
				m.add(n);
			}
		}
		Collections.sort(p);
		Collections.sort(m);
		
		long ans = 0;
		while(p.size() >= 2) {
			long p1 = p.remove(p.size()-1);
			long p2 = p.remove(p.size()-1);
			if (p1 == 1 || p2 == 1) {
				ans += p1 + p2;
			}
			else {
				ans += p1 *p2;
			}
		}
		if (p.size() == 1) {
			long p3 = p.remove(p.size()-1);
			ans += p3;
		}
		
		while(m.size() >= 2) {
			long m1 = m.remove(0);
			long m2 = m.remove(0);
			ans += m1 *m2;
		}
		if (m.size() == 1) {
			long m3 = m.remove(0);
			ans += m3;
		}
		System.out.println(ans);
	}
}